src/jit/lsra.cpp

```diff
@@ -4857,15 +4901,40 @@ LinearScan::tryAllocateFreeReg(Interval *currentInterval, RefPosition *refPositi
 //                  and that can be spilled.
 //
 // Arguments:
-//    current:         The interval for the current allocation
-//    refPosition:     The RefPosition of the current Interval for which a register is being allocated
+//    current               The interval for the current allocation
+//    refPosition           The RefPosition of the current Interval for which a register is being allocated
+//    allocationOptional    If true, a reg may not be allocated if all other ref positions currently
+//                          occupying registers are more important than the 'refPosition'.
 //
 // Return Value:
 //    The regNumber allocated to the RefPositon.  Returns REG_NA if no free register is found.
-
+//
+// Note:  Currently this routine uses weight and farthest distance of next reference
+// to select a ref position for spilling.  
+// a) if allocationOptional = false
+//        The ref position chosen for spilling will be the lowest weight
+//        of all and if there is is more than one ref position with the
+//        same lowest weight, among them choses the one with farthest
+//        distance to its next reference.
+// 
+// b) if allocationOptional = true
+//        The ref position chosen for spilling will not only be lowest weight
+//        of all but also has a weight lower than 'refPosition'.  If there is
+//        no such ref position, reg will not be allocated.
 regNumber
-LinearScan::allocateBusyReg(Interval *current, RefPosition *refPosition)
+LinearScan::allocateBusyReg(Interval* current,
+                            RefPosition* refPosition,
+                            bool allocationOptional)
 {
```

Comment:
Right now allocateBusyReg() is called from two places:

1) allocateRegisters() - at this place whether reg is optional is comes from the ref position.
2) assignCopyReg() - at this place we must need a reg allocated by spilling another interval if required, even if the ref position is marked as reg optional.  Because it could be possible that we allocated a reg to reg-optional refposition because a free register is available.

As you can see from above, allocateBusyReg() cannot always use refPosition->RequiresRegOptionally() because in one code path it must require a register irrespective of refPosition is reg optional or not.  For that reason I had to add allocationOptional parameter.